### PR TITLE
Implements logs subcommand

### LINKED ISSUE #208: Implement logs subcommand

### PR DESCRIPTION
**Pull request by chuckha**
Closes #208

Signed-off-by: Chuck Ha <chuck@heptio.com>

### CODE REVIEW

File: cmd/sonobuoy/app/logs.go [hunk 1]
```diff
@@ -31,15 +35,55 @@ func init() {
...
+		Follow: follow,
+	}
+	if err := ops.GetLogs(clientset, logOptions); err != nil {
 		errlog.LogError(errors.Wrap(err, "error attempting to get sonobuoy logs"))
-		code = 1
+		os.Exit(1)
+	}
+}
+
+func getClientset() (*kubernetes.Clientset, error) {
```
Comment (timothysc) [inline cmd/sonobuoy/app/logs.go]: This probably belongs in config as a more general f(n) b/c we do similar things in multiple locations.

File: cmd/sonobuoy/app/logs.go [hunk 2]
```diff
@@ -31,15 +35,55 @@ func init() {
...
+	}
+
+	follow, err := cmd.Flags().GetBool("follow")
+	if err != nil {
+		errlog.LogError(errors.Wrap(err, "error getting follow flag"))
+		os.Exit(1)
+	}
+
+	logOptions := &ops.LogOptions{
+		Follow: follow,
```
Comment (timothysc) [inline cmd/sonobuoy/app/logs.go]: If the number of options over time still = 1 then we just flatten this.

File: cmd/sonobuoy/app/operations/logs.go [hunk 3]
```diff
@@ -17,11 +17,62 @@ limitations under the License.
...
 )
 
-// Logs gathers the logs for the containers in the sonobuoy namespace and prints them out in a structured format
-func GetLogs( /*opts?*/ ) error {
-	// This should be straight forwards
-	return errors.New("not implemented")
+// Logs gathers the logs for the containers in the sonobuoy namespace and prints them
+
+// TODO(chuckha) should this be an option?
+const sonobuoyNamespace = "heptio-sonobuoy"
```
Comment (timothysc) [inline cmd/sonobuoy/app/operations/logs.go]: It is pulled in from config, we should leverage the init routine to grab that data.

File: cmd/sonobuoy/app/operations/logs.go [hunk 4]
```diff
@@ -17,11 +17,62 @@ limitations under the License.
...
+	}
+
+	// Otherwise print every log we can
+	pods, err := client.CoreV1().Pods(sonobuoyNamespace).List(metav1.ListOptions{})
+	if err != nil {
+		return errors.Wrap(err, "could not list pods")
+	}
+	for _, pod := range pods.Items {
+		for _, container := range pod.Spec.Containers {
+			logrus.WithFields(logrus.Fields{"pod": pod.Name, "container": container.Name}).Info("Printing container logs")
```
Comment (timothysc) [inline cmd/sonobuoy/app/operations/logs.go]: You will probably some larger separator for containers.

File: cmd/sonobuoy/app/operations/logs.go [hunk 5]
```diff
@@ -17,11 +17,62 @@ limitations under the License.
...
+			}
+		}
+	}
+	return nil
+}
+
+// TODO(chuckha) the output is a little confusing because our containers already produce structured logs.
+
+func streamLogs(client kubernetes.Interface, podName string, logOptions *v1.PodLogOptions) error {
+	req := client.CoreV1().Pods(sonobuoyNamespace).GetLogs("sonobuoy", logOptions)
```
Comment (timothysc) [inline cmd/sonobuoy/app/operations/logs.go]: const "sonobuoy" from Config.

File: cmd/sonobuoy/app/logs.go [hunk 6]
```diff
@@ -31,15 +35,55 @@ func init() {
...
 	}
-	// TODO: Determine options
+	cmd.Flags().BoolP("follow", "f", false, "Specify if the logs should be streamed.")
 	RootCmd.AddCommand(cmd)
 }
 
 func getLogs(cmd *cobra.Command, args []string) {
-	code := 0
-	if err := ops.GetLogs( /*opts*/ ); err != nil {
+	clientset, err := getClientset()
```
Comment (timothysc) [inline cmd/sonobuoy/app/logs.go]: we could build a singleton utility function in config that does this.

File: cmd/sonobuoy/app/logs.go (hunk 2 above)
Comment (chuckha) [reply cmd/sonobuoy/app/logs.go]: flattened because we can expand it later.

File: cmd/sonobuoy/app/logs.go (hunk 6 above)
Comment (chuckha) [reply cmd/sonobuoy/app/logs.go]: I think I can make LoadClient do a little bit more work and we can use that most places.

File: cmd/sonobuoy/app/operations/logs.go [hunk 7]
```diff
@@ -17,11 +17,66 @@ limitations under the License.
...
+	for _, pod := range pods.Items {
+		for _, container := range pod.Spec.Containers {
+			logrus.WithFields(logrus.Fields{"pod": pod.Name, "container": container.Name}).Info("Printing container logs")
+			err := streamLogs(client, cfg.PluginNamespace, pod.Name, &v1.PodLogOptions{
+				Container: container.Name,
+			})
+			if err != nil {
+				return errors.Wrap(err, "failed to stream logs")
+			}
+			fmt.Println(podLogSeparator)
```
Comment (chuckha) [inline cmd/sonobuoy/app/operations/logs.go]: :/ not happy about this. Think it's alright?

Comment (chuckha) [reply cmd/sonobuoy/app/operations/logs.go]: I was also thinking of disabling this line with -q but maybe we can implement that later if need be.

File: pkg/config/loader.go [hunk 8]
```diff
@@ -147,6 +149,39 @@ func LoadClient(cfg *Config) (kubernetes.Interface, error) {
...
+		return nil, errors.Wrap(err, "could not build config from kubeconfig")
+	}
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not make a new clientset from config")
+	}
+	return clientset, nil
+}
+
+func locateKubeconfig() string {
```
Comment (liztio) [inline pkg/config/loader.go]: Could we use Viper for this maybe? Might be nice to be able to pass in --config or something

File: cmd/sonobuoy/app/operations/logs.go [hunk 9]
```diff
@@ -17,11 +17,66 @@ limitations under the License.
...
 )
 
-// Logs gathers the logs for the containers in the sonobuoy namespace and prints them out in a structured format
-func GetLogs( /*opts?*/ ) error {
-	// This should be straight forwards
-	return errors.New("not implemented")
+// Logs gathers the logs for the containers in the sonobuoy namespace and prints them
+
+// GetLogs streams logs from the sonobuoy pod by default to stdout.
+func GetLogs(cfg *config.Config, follow bool) error {
```
Comment (liztio) [inline cmd/sonobuoy/app/operations/logs.go]: could `config.Config` be a bit narrower? Seems like a big thunk to pass in to a single log function

File: pkg/config/loader.go [hunk 10]
```diff
@@ -147,6 +149,39 @@ func LoadClient(cfg *Config) (kubernetes.Interface, error) {
 	return clientset, nil
 }
 
+func OutOfClusterClient() (kubernetes.Interface, error) {
```
Comment (liztio) [inline pkg/config/loader.go]: This should probably have a comment. What does `OutOfClusterClient` mean?

File: cmd/sonobuoy/app/operations/logs.go (hunk 9 above)
Comment (chuckha) [reply cmd/sonobuoy/app/operations/logs.go]: yeah it is, i only need the namespace from the config so no reason to pass it in at all

File: pkg/config/loader.go (hunk 10 above)
Comment (chuckha) [reply pkg/config/loader.go]: I added a comment, the language came from https://github.com/kubernetes/client-go/tree/master/examples 

in cluster is stuff running in the cluster, like sonobuoy https://github.com/heptio/sonobuoy/blob/master/pkg/config/loader.go#L135

and similarly out of cluster is stuff accessing the cluster from outside

Comment (liztio) [reply pkg/config/loader.go]: that makes sense! I don't love that it doesn't take any arguments, though. Could it take a kubeconfig filename?

File: cmd/sonobuoy/app/operations/logs.go (hunk 7 above)
Comment (timothysc) [reply cmd/sonobuoy/app/operations/logs.go]: I'm ok with it.  Eventually you may want a utility f(n) like prettyPrint for a container with an optional separator arg.

File: pkg/config/loader.go (hunk 10 above)
Comment (timothysc) [inline pkg/config/loader.go]: I believe you can conditionalize to detect in cluster and fallback and we can use in any location then.  

Also //Comment Exported.

File: cmd/sonobuoy/app/constants/constants.go [hunk 11]
```diff
@@ -0,0 +1,5 @@
+package constants
+
+const (
+	SonobuoyPod = "sonobuoy"
```
[the remark this answers is not in the document]
Comment (chuckha) [reply cmd/sonobuoy/app/constants/constants.go]: yeah true, i'll collapse now and we can split out later when we need it.

**Pull request merged**